Improve documentation about Test Categories + Fix some code to make it consistent with the doc

iemejia opened the pull request:
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [ ] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [ ] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

iemejia commented:
R: @kennknowles

iemejia commented:
R: @dhalperi since we briefly discussed about this.

Notice that my changes to the BigQuery and DataStore IOs can have unintended consequences, so please guide me in case you think I should not change those.

dhalperi commented:
R: @jasonkuster want to TAL?

dhalperi commented on sdks/java/io/elasticsearch/src/test/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIOTest.java:
```diff
@@ -135,7 +135,7 @@ public void testSizes() throws Exception {
 
   @Test
-  @Category(RunnableOnService.class)
+  @Category(NeedsRunner.class)
```
just drop altogether?

jasonkuster commented:
:) Thanks Dan, happy to.

dhalperi commented on sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOTest.java:
```diff
@@ -1308,8 +1307,8 @@ public void testBuildSourceDisplayData() {
   @Test
-  @Category(RunnableOnService.class)
-  @Ignore("[BEAM-436] DirectRunner RunnableOnService tempLocation configuration insufficient")
+  @Category(NeedsRunner.class)
```
RunnableOnService is desired here. We want to check that runners preserve the display data.

iemejia replied:
ok

dhalperi commented on sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableIOTest.java:
```diff
@@ -285,6 +285,7 @@ public void testReadingFailsTableDoesNotExist() throws Exception {
 
   /** Tests that when reading from an empty table, the read succeeds. */
   @Test
+  @Category(NeedsRunner.class)
```
why add?

iemejia replied:
Because what I understood of annotating this tests is to 'tag' them as tests that use a runner to be executed. In addition the nodeEnforcement changes that stas did get enforced e.g. for TestPipelines with transforms that don't call run().

iemejia replied:
And to be coherent with the other IOs.

dhalperi commented on sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/datastore/DatastoreV1Test.java:
```diff
@@ -364,7 +364,7 @@ public void testDeleteEntityPrimitiveDisplayData() {
 
   @Test
-  @Category(RunnableOnService.class)
+  @Category(NeedsRunner.class)
```
revert? same as BigQueryIO.

iemejia replied:
ok

dhalperi commented on sdks/java/io/mqtt/src/test/java/org/apache/beam/sdk/io/mqtt/MqttIOTest.java:
```diff
@@ -78,7 +78,7 @@ public void startBroker() throws Exception {
 
   @Test(timeout = 60 * 1000)
-  @Category(RunnableOnService.class)
+  @Category(NeedsRunner.class)
```
drop altogether?

iemejia replied:
Same as before, I think there is some value on tagging the tests that use a runner, to enforce validations.

iemejia replied:
And also this is done in all the other IOs, so I am trying to get things to be similar..

kennknowles commented:
R: -@kennknowles

I'll bow out as there is already good attention here.

iemejia commented:
Rebased (I just removed the changes for the GCP IOs and updated to most recent master). Please check if it is ok, or there are other things to improve.

dhalperi commented:
My belief: `NeedsRunner` should only be used in modules that are depended on  (directly or transitively) by the `DirectRunner`. In all other cases, the modules can simply depend on the `DirectRunner` in test scope and the category is useless.

@jasonkuster @davorbonaci ?

iemejia commented:
Mmmm... that adds another idea to the list, well, the good thing is that at the end we will finally know the exact goals, so if it is as you said then all the tests in the IOs that use a runner should be tagged with RunnableOnService ?

iemejia commented:
@kennknowles (sorry to ping you again), but I have the impression that what Dan said, opens another dimension, or I didn't understand when we discussed on slack then.

davorbonaci commented:
+1 on @dhalperi's comment on the usage of `@NeedsRunner` annotation.

iemejia commented:
Ok guys, I understand the logic and in that case Dan's recommendation to remove the annotation makes total sense. One question, this means that not any single IO should have NeedsRunner in their tests ?
I am seeing that BigQueryIOTest has them, so should I remove those too ?

iemejia commented:
On extra thing that I forgot, I see some value in tagging tests to indicate that they require a runner because this will trigger the nodeEnforcement validations that @staslev put in place, those won't be executed if I remove the NeedsRunner annotation. Other idea would be to have a different annotation to enforce such cases.
Please tell me what do you prefer and I will do it.

staslev commented:
I have put some info in writing, which I hope this discussion can find useful.
Please feel free to correct me if I got things wrong.

_What are the `@NeedsRunner` and `@RunnableOnService`annotations needed for_?

* `@NeedsRunner` is a way for tests residing in `beam-sdks-java-core` / `beam-runners-core-java` to request the `DirectRunner` to execute them as part of its build cycle.
* `@RunnableOnService` is a way for tests in `beam-sdks-java-core` to request the other runners (e.g. `Spark`, `Flink` runners) to execute them as part of their build cycle. 

The `DirectRunner` is set to specifically search for `@NeedsRunner` tests in `beam-sdks-java-core` and `beam-runners-core-java` during its `integration-test` phase and run then in addition to its own tests. 
```xml
<dependenciesToScan>
  <dependency>org.apache.beam:beam-sdks-java-core</dependency>
  <dependency>org.apache.beam:beam-runners-core-java</dependency>
</dependenciesToScan>
```
Similarly, `@RunnableOnService` tests are searched for and executed by the other runners during the same phase.

```xml
<dependenciesToScan>
  <dependency>org.apache.beam:beam-sdks-java-core</dependency>
</dependenciesToScan>
```

Further aspects in play are the `TestPipeline` JUnit rule,  the `beamTestPipelineOptions`, and `beamUseDummyRunner` system properties. When the `TestPipeline` is used in the scope of a test, it checks the `beamTestPipelineOptions` to determine the pipeline options, among which is the runner class, so each runner module can set its own test settings using the `beamTestPipelineOptions`. 
If `beamTestPipelineOptions` is not set, the `beamUseDummyRunner` system property is inspected, and if set, a special type of runner is set - the `CrashingRunner`, which throws an `IllegalArgumentException` exception upon an attempt to execute the pipeline. `CrashingRunner` is suitable for tests that test pipeline construction only, and do not actually run the pipeline they construct.

_Do tests outside `beam-sdks-java-core` / `beam-runners-core-java` need to be decorated with `@NeedsRunner`/`@RunnableOnService`_?

`DirectRunner` does NOT search for tests annotated with `@NeedsRunner` outside the `beam-sdks-java-core` and `beam-runners-core-java` modules. So even if they are annotated as such `@NeedsRunner` they will NOT be executed as part of the `DirectRunner`s build cycle. 

However, `TestPipeline` offers mechanisms to avoid common pitfalls when designing pipeline tests,  mechanisms that are activated by detecting a `@NeedsRunner` / `@RunnableOnService` annotation on a test. Thus it may be worthwhile adding a `@NeedsRunner` so that these mechanisms are activated and can do their job.

Alternatively, `@NeedsRunner` may be omitted, and these mechanisms can be activated manually by setting `testPipeline.enableAbandonedNodeEnforcement(true)` as part of the test.

iemejia commented:
Wow excellent @staslev. This is super clear (and totally deserves a PR to add this to the testing guide). I will add the appropriate information to the javadoc in this PR thanks.

So assuming that you (the others) agree with Stas, what is the next step (for the case of these annotations in the IOs [The case that i care)?  some options:
1. Create a new category to indicate that the running of the tests should be validated using the TestPipeline validations ?
2. Keep  NeedsRunner in the IOs like it is and have this mixed behavior (not really called, but enforced by TestPipeline too).
3. Get rid of all the Categories in the IOs since they are not really enforced (again asuming what Stas wrote) ?

WDYT?

dhalperi commented:
That is indeed pretty cool.

Stas, I'm curious about using `@NeedsRunner` to decide whether the pipeline should be run. Couldn't you instead check whether the pipeline had anything applied? Or have users simply opt-out by default?

staslev commented:
Just to clarify this a bit, `@NeedsRunner` is used by the `TestPipeline` to decide whether to activate the abandoned node detection, not whether to run the pipeline.

I think it boils down to `@NeedsRunner`/`@RunnableOnService` being a bit confusing outside the context of the sdk modules, which brings us back to the question of whether they should or should not be used outside of the `beam-sdks-java-core / beam-runners-core-java` modules. `TestPipeline` can be then adjusted accordingly.

kennknowles commented:
R: -@kennknowles

I'm late to this, assuming it is OK to remove myself because it looks really well handled. Feel free to re-add me if you want - or perhaps some of this should be brought to the dev list?

iemejia commented:
The goal of these PR is basically to improve the doc and to fix the correct use of these tests categories in the existing IOs.
The first part I think is already clear and I will improve the docs taking the ideas from Stas comments. The second part however is still not clear for me, what do you think? should the IOs don't have any category annotation? or which of the three options I presented before makes sense?

iemejia commented:
And just for ref...
https://github.com/apache/beam/pull/2157

kennknowles commented:
One thing I would request, if `@NeedsRunner` is omitted so that it is expected that `DirectRunner` is the only runner that executes a test in a non-core module, it would be better for this to be explicit via setting up the options than "magic" of having that happen just because it is the only runner on the classpath.

kennknowles commented on sdks/java/core/src/main/java/org/apache/beam/sdk/testing/NeedsRunner.java:
```diff
@@ -21,8 +21,23 @@
- * Category tag for validation tests which utilize {@link TestPipeline} for execution and expect
- * to be executed by a {@link PipelineRunner}.
+ * Category tag for tests which utilize {@link TestPipeline} for execution and
+ * {@link PAssert} for validation. Tests in this category require a
```
It is not necessary that they use `PAssert`. It just means that "this test cannot run without a runner set up for `TestPipeline`".

kennknowles commented on sdks/java/core/src/main/java/org/apache/beam/sdk/testing/RunnableOnService.java:
```diff
@@ -17,9 +17,12 @@
 /**
- * Category tag for validation tests which utilize {@link TestPipeline} for execution and
- * {@link PAssert} for validation. Example usage:
+ * Category tag for tests which utilize {@link TestPipeline} for execution and
```
This will now be `ValidatesRunner` and I think it should say something about the fact that "this is a test of the runner itself, not (necessarily) the transforms being executed by the runner". Exact wording TBD, and there's the outstanding question of what that means for IO ITs.

kennknowles commented on sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableIOTest.java:
```diff
@@ -296,6 +297,7 @@ public void testReadingEmptyTable() throws Exception {
 
   /** Tests reading all rows from a table. */
   @Test
+  @Category(NeedsRunner.class)
```
I'm slightly on the fence about these. I think this is an accurate use of the category, yet I feel like there might be another interesting subcategory that we should create. Let's move the discussion to dev list and come up with a solution there.
